File: game.go
package main

import (
	"math"
	"os"

	rl "github.com/gen2brain/raylib-go/raylib"

	"github.com/njayman/galaxyimpact/highscore"
)

type GameState int

const (
	TITLE GameState = iota
	GAMEPLAY
	PAUSED
	GAME_OVER
	LEVEL_UP
	SETTINGS
)

// arenaHalf bounds the otherwise-open world: play reads as infinite (no
// visible edges under normal movement) but is backed by this large finite
// square so spawn-ring/background math stays simple bounded arithmetic.
const arenaHalf = float32(4000)

// waveDuration is how long (seconds) each wave lasts before the next begins.
const waveDuration = float32(25)

const highScoreFile = "highscores.txt"

type Game struct {
	// ScreenWidth/ScreenHeight are the fixed logical design resolution that all
	// game logic and drawing coordinates are expressed in - never changed at
	// runtime. WindowWidth/WindowHeight track the actual OS window size (which
	// does change, e.g. in fullscreen) purely so the final frame can be scaled
	// and letterboxed to fit it without anything in the game moving or resizing.
	ScreenWidth             int32
	ScreenHeight            int32
	WindowWidth             int32
	WindowHeight            int32
	State                   GameState
	Player                  Player
	Boss                    Boss
	BossActive              bool
	Bullets                 []Bullet
	Asteroids               []Asteroid
	Projectiles             []BossProjectile
	Enemies                 []Enemy
	Pickups                 []Pickup
	Mines                   []Mine
	Weapons                 []Weapon
	SkillLevels             map[SkillID]int32
	PendingChoices          []LevelUpChoice
	PostCapDamageLevels     int32
	BossDeathShockwave      bool
	BossDeathShockwaveTimer float32
	BossDeathShockwavePos   rl.Vector2
	BossDeathShockwaveHit   bool
	BlackHole               BlackHole
	Stars                   []Star
	BgParticles             []BgParticle
	BorderStars             []Star
	BoundaryClouds          []GasCloud
	DeathParticles          []Particle
	AsteroidSpawnTimer      float32
	EnemySpawnTimer         float32
	SpreadWindupShots       int32
	XP                      int32
	Level                   int32
	XPToNext                int32
	WaveNumber              int32
	WaveTimer               float32
	RunTime                 float32
	Score                   int32
	HighScores              []int32
	HighScoreRepo           highscore.Repository
	MenuIndex               int32
	ScoreRecorded           bool
	Sounds                  Sounds
	ShakeTimer              float32
	ShakeDuration           float32
	ShakeIntensity          float32
	HitPauseTimer           float32
	WorldTarget             rl.RenderTexture2D
	PixelTarget             rl.RenderTexture2D
	Font                    rl.Font
	Settings                Settings
	SettingsReturnState     GameState
	BGM                     rl.Music
	Sandbox                 bool
	SandboxKindIndex        int32
}

// pixelScale downscales WorldTarget (game-world shapes only: ship, enemies,
// bullets, background, ...) before it's scaled back up with nearest-neighbor
// filtering onto PixelTarget - this is what gives the deliberate chunky
// pixel-art look. Text is drawn separately, straight onto PixelTarget at its
// native ScreenWidth/ScreenHeight resolution (see DrawGame), so it stays
// crisp instead of inheriting the same blur a low-res-then-upscaled font
// would have. The final PixelTarget->window blit (see letterboxRect) is a
// separate, resolution-independence concern, unrelated to this chunkiness.
const pixelScale = int32(3)

// defaultWindowWidth/Height is the default logical resolution and starting
// window size - Full HD, so text/HUD have plenty of native pixels to work
// with even before any fullscreen/letterbox scaling.
const (
	defaultWindowWidth  = int32(1920)
	defaultWindowHeight = int32(1080)
)

func InitGame() *Game {
	g := &Game{}

	g.ScreenWidth = defaultWindowWidth
	g.ScreenHeight = defaultWindowHeight
	g.WindowWidth = defaultWindowWidth
	g.WindowHeight = defaultWindowHeight
	g.State = TITLE

	g.Stars = make([]Star, 80)
	for i := range g.Stars {
		g.Stars[i] = Star{
			Position: rl.NewVector2(float32(rl.GetRandomValue(0, g.ScreenWidth)), float32(rl.GetRandomValue(0, g.ScreenHeight))),
			Radius:   float32(rl.GetRandomValue(10, 20)) / 10.0,
		}
	}

	bgColors := []rl.Color{rl.Fade(colorStructMid, 0.15), rl.Fade(colorHaze, 0.1), rl.Fade(colorAccentDim, 0.08)}
	g.BgParticles = make([]BgParticle, 25)
	for i := range g.BgParticles {
		g.BgParticles[i] = BgParticle{
			Position: rl.NewVector2(float32(rl.GetRandomValue(0, g.ScreenWidth)), float32(rl.GetRandomValue(0, g.ScreenHeight))),
			Velocity: rl.NewVector2(float32(rl.GetRandomValue(-4, 4))/10.0, float32(rl.GetRandomValue(-4, 4))/10.0),
			Radius:   float32(rl.GetRandomValue(15, 40)) / 10.0,
			Color:    bgColors[rl.GetRandomValue(0, int32(len(bgColors)-1))],
		}
	}

	// BorderStars cover a large virtual area (up to 8K) at native resolution so
	// the letterbox bars around the pixel-art frame show more starfield instead
	// of flat black, on any window/monitor size.
	g.BorderStars = make([]Star, 400)
	for i := range g.BorderStars {
		g.BorderStars[i] = Star{
			Position: rl.NewVector2(float32(rl.GetRandomValue(0, 7680)), float32(rl.GetRandomValue(0, 4320))),
			Radius:   float32(rl.GetRandomValue(10, 20)) / 10.0,
		}
	}

	g.BoundaryClouds = generateBoundaryClouds()

	g.HighScoreRepo = highscore.NewFileRepository(highScoreFile)
	g.HighScores, _ = g.HighScoreRepo.Load()

	g.Sounds = LoadSounds()
	g.Font = loadReadableFont()

	g.Settings = loadSettings()

	g.BGM = loadBGM()
	rl.PlayMusicStream(g.BGM)

	g.WorldTarget = rl.LoadRenderTexture(g.ScreenWidth/pixelScale, g.ScreenHeight/pixelScale)
	rl.SetTextureFilter(g.WorldTarget.Texture, rl.FilterPoint)

	g.PixelTarget = rl.LoadRenderTexture(g.ScreenWidth, g.ScreenHeight)
	rl.SetTextureFilter(g.PixelTarget.Texture, rl.FilterPoint)

	resetRun(g)
	g.State = TITLE

	return g
}

// generateBoundaryClouds builds an irregular nebula ring around arenaHalf -
// clusters of overlapping soft blobs, jittered in angle/distance/size, so
// the world's edge reads as an organic gas cloud rather than a perfect
// circle or an invisible wall. The actual clamp is still just a circular
// distance check (see updatePlayerMovement) - this is decoration only.
func generateBoundaryClouds() []GasCloud {
	cloudColors := []rl.Color{rl.Fade(colorHaze, 0.12), rl.Fade(colorStructMid, 0.15), rl.Fade(colorAccentDim, 0.08)}

	var clouds []GasCloud
	const clusterCount = 40

	for i := 0; i < clusterCount; i++ {
		baseAngle := float64(i) * 360.0 / clusterCount
		angle := (baseAngle + float64(rl.GetRandomValue(-8, 8))) * rl.Deg2rad
		dist := arenaHalf + float32(rl.GetRandomValue(-800, 800))
		center := rl.NewVector2(float32(math.Cos(angle))*dist, float32(math.Sin(angle))*dist)

		blobCount := int(rl.GetRandomValue(3, 5))
		for b := 0; b < blobCount; b++ {
			offset := rl.NewVector2(float32(rl.GetRandomValue(-150, 150)), float32(rl.GetRandomValue(-150, 150)))
			clouds = append(clouds, GasCloud{
				Position: rl.Vector2Add(center, offset),
				Radius:   float32(rl.GetRandomValue(80, 220)),
				Color:    cloudColors[rl.GetRandomValue(0, int32(len(cloudColors)-1))],
			})
		}
	}

	return clouds
}

// loadReadableFont picks a real, legible system sans-serif over raylib's tiny
// built-in bitmap font (which is what made text unreadable). Loaded at a
// large base size and downsampled at draw time for crisp text at any size.
// Falls back to the default font if none of the common paths exist.
func loadReadableFont() rl.Font {
	candidates := []string{
		"/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf",
		"/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf",
		"/usr/share/fonts/truetype/noto/NotoSans-Bold.ttf",
		"/Library/Fonts/Arial Bold.ttf",
		"C:\\Windows\\Fonts\\arialbd.ttf",
	}

	for _, path := range candidates {
		if _, err := os.Stat(path); err == nil {
			font := rl.LoadFontEx(path, 96, nil)
			if font.Texture.ID != 0 {
				rl.SetTextureFilter(font.Texture, rl.FilterBilinear)
				return font
			}
		}
	}

	return rl.GetFontDefault()
}

// toggleFullscreen switches between windowed and fullscreen. Going fullscreen
// resizes the window to the current monitor's native resolution first (so it
// fills displays up to 8K), then hands off to raylib's fullscreen mode. The
// game's own logical resolution never changes - see syncScreenSize.
func toggleFullscreen(g *Game) {
	if rl.IsWindowFullscreen() {
		rl.ToggleFullscreen()
		rl.SetWindowSize(int(defaultWindowWidth), int(defaultWindowHeight))
	} else {
		monitor := rl.GetCurrentMonitor()
		rl.SetWindowSize(rl.GetMonitorWidth(monitor), rl.GetMonitorHeight(monitor))
		rl.ToggleFullscreen()
	}

	syncScreenSize(g)
}

// syncScreenSize tracks the actual window size (whether it changed via
// toggleFullscreen or the player dragging the resizable window's edges) so
// DrawGame can scale+letterbox the fixed-resolution frame to fit it. Nothing
// about the game's own coordinate space or render target changes here - that
// fixed logical resolution is what keeps every element, text included,
// scaling uniformly instead of shrinking relative to a bigger window.
func syncScreenSize(g *Game) {
	g.WindowWidth = int32(rl.GetScreenWidth())
	g.WindowHeight = int32(rl.GetScreenHeight())
}

// resetRun (re)initializes everything needed for a fresh playthrough, without
// touching window-level state (screen size, starfield, loaded high scores).
// The player always starts at the world origin - the camera keeps them
// pinned at screen center regardless (see beginWorldCamera in draw.go).
func resetRun(g *Game) {
	g.Player = Player{
		Position:  rl.Vector2{},
		Radius:    15,
		Color:     colorAccent,
		Speed:     5,
		Health:    5,
		MaxHealth: 5,
		Charges:   maxCharges,
	}

	g.Boss = Boss{Size: rl.NewVector2(100, 100)}
	g.BossActive = false

	g.Bullets = []Bullet{}
	g.Asteroids = []Asteroid{}
	g.Projectiles = []BossProjectile{}
	g.Enemies = []Enemy{}
	g.Pickups = []Pickup{}
	g.Mines = []Mine{}
	g.DeathParticles = []Particle{}
	g.Weapons = []Weapon{{Kind: WeaponForward, Level: 1}}
	g.SkillLevels = map[SkillID]int32{SkillForwardShot: 1}
	g.PostCapDamageLevels = 0
	g.BossDeathShockwave = false
	g.BossDeathShockwaveTimer = 0
	g.BossDeathShockwaveHit = false

	g.BlackHole = BlackHole{Timer: float32(rl.GetRandomValue(30, 60)) / 10.0}
	g.AsteroidSpawnTimer = 1.0
	g.EnemySpawnTimer = 1.0
	g.SpreadWindupShots = 0

	g.XP = 0
	g.Level = 1
	g.XPToNext = 100
	g.WaveNumber = 1
	g.WaveTimer = waveDuration
	g.RunTime = 0

	g.ShakeTimer = 0
	g.ShakeDuration = 0
	g.ShakeIntensity = 0
	g.HitPauseTimer = 0
	g.Score = 0
	g.ScoreRecorded = false
	g.State = GAMEPLAY
}

File: src/game.cpp
#include "game.hpp"

#include "entities/ship.hpp"
#include "highscore.hpp"
#include "palette.hpp"
#include "platform.hpp"
#include "raymath.h"
#include "sound.hpp"
#include "update.hpp"
#include "update_constants.hpp"
#include <algorithm>
#include <array>
#include <cmath>
#include <filesystem>
#include <memory>

namespace
{

auto generateBoundaryClouds() -> std::vector<GasCloud>
{
    const std::array<Color, 3> cloudColors{Fade(Palette::Haze, 0.12F),
                                           Fade(Palette::StructMid, 0.15F),
                                           Fade(Palette::AccentDim, 0.08F)};

    std::vector<GasCloud> clouds;
    constexpr int clusterCount = 40;

    for (int i = 0; i < clusterCount; i++)
    {
        const float baseAngle = static_cast<float>(i) * 360.0F / clusterCount;
        const float angle = (baseAngle + static_cast<float>(GetRandomValue(-8, 8))) * DEG2RAD;
        const float dist = GameConstants::arenaHalf + static_cast<float>(GetRandomValue(-800, 800));
        const Vector2 center{.x = std::cos(angle) * dist, .y = std::sin(angle) * dist};

        const int blobCount = GetRandomValue(3, 5);
        for (int b = 0; b < blobCount; b++)
        {
            const Vector2 offset{.x = static_cast<float>(GetRandomValue(-150, 150)),
                                 .y = static_cast<float>(GetRandomValue(-150, 150))};
            clouds.push_back(GasCloud{.position = Vector2Add(center, offset),
                                      .radius = static_cast<float>(GetRandomValue(80, 220)),
                                      .color = cloudColors.at(static_cast<size_t>(GetRandomValue(
                                          0, static_cast<int32_t>(cloudColors.size() - 1))))});
        }
    }

    return clouds;
}

auto loadGameFont() -> Font
{
    const std::string path =
        std::string(GetApplicationDirectory()) + "assets/fonts/HeavyDataNerdFont-Regular.ttf";

    if (std::filesystem::exists(path))
    {
        const Font font = LoadFontEx(path.c_str(), 96, nullptr, 0);
        if (font.texture.id != 0)
        {
            SetTextureFilter(font.texture, TEXTURE_FILTER_BILINEAR);
            return font;
        }
    }

    return GetFontDefault();
}

auto computeRenderScale(const Game& game) -> float
{
    const float widthScale = static_cast<float>(game.resources.windowWidth) /
                             static_cast<float>(game.resources.screenWidth);
    const float heightScale = static_cast<float>(game.resources.windowHeight) /
                              static_cast<float>(game.resources.screenHeight);
    return std::min(widthScale, heightScale);
}

void createRenderTargets(Game& game)
{
    const auto width = std::max(1, static_cast<int>(static_cast<float>(game.resources.screenWidth) *
                                                    game.resources.renderScale));
    const auto height =
        std::max(1, static_cast<int>(static_cast<float>(game.resources.screenHeight) *
                                     game.resources.renderScale));

    game.resources.worldTarget.load(width, height);
    SetTextureFilter(game.resources.worldTarget.get().texture, TEXTURE_FILTER_BILINEAR);

    game.resources.pixelTarget.load(width, height);
    SetTextureFilter(game.resources.pixelTarget.get().texture, TEXTURE_FILTER_BILINEAR);
}

}

auto InitGame() -> Game
{
    Game game{};

    game.resources.screenWidth = GameConstants::defaultWindowWidth;
    game.resources.screenHeight = GameConstants::defaultWindowHeight;
    game.resources.windowWidth = GameConstants::defaultWindowWidth;
    game.resources.windowHeight = GameConstants::defaultWindowHeight;
    game.state = GameState::TITLE;

    game.run.stars.resize(80);
    for (auto& star : game.run.stars)
    {
        star = Star{
            .position =
                Vector2{.x = static_cast<float>(GetRandomValue(0, game.resources.screenWidth)),
                        .y = static_cast<float>(GetRandomValue(0, game.resources.screenHeight))},
            .radius = static_cast<float>(GetRandomValue(10, 20)) / 10.0F};
    }

    const std::array<Color, 3> bgColors{Fade(Palette::StructMid, 0.15F), Fade(Palette::Haze, 0.1F),
                                        Fade(Palette::AccentDim, 0.08F)};
    game.run.bgParticles.resize(25);
    for (auto& particle : game.run.bgParticles)
    {
        particle = BgParticle{
            .position =
                Vector2{.x = static_cast<float>(GetRandomValue(0, game.resources.screenWidth)),
                        .y = static_cast<float>(GetRandomValue(0, game.resources.screenHeight))},
            .velocity = Vector2{.x = static_cast<float>(GetRandomValue(-4, 4)) / 10.0F,
                                .y = static_cast<float>(GetRandomValue(-4, 4)) / 10.0F},
            .radius = static_cast<float>(GetRandomValue(15, 40)) / 10.0F,
            .color = bgColors.at(
                static_cast<size_t>(GetRandomValue(0, static_cast<int32_t>(bgColors.size() - 1))))};
    }

    game.run.borderStars.resize(400);
    for (auto& star : game.run.borderStars)
    {
        star = Star{.position = Vector2{.x = static_cast<float>(GetRandomValue(0, 7680)),
                                        .y = static_cast<float>(GetRandomValue(0, 4320))},
                    .radius = static_cast<float>(GetRandomValue(10, 20)) / 10.0F};
    }

    game.run.gasClouds = generateBoundaryClouds();

    game.resources.highScoreRepo = std::make_shared<highscore::FileRepository>(
        getSaveDataDir() + GameConstants::highScoreFile);
    game.resources.highScores = game.resources.highScoreRepo->load();

    game.resources.sounds = LoadSounds();
    game.resources.font = loadGameFont();

    game.resources.settings = loadSettings();
    game.resources.achievements = loadAchievements();

    game.resources.bgm = loadBGM();
    PlayMusicStream(game.resources.bgm.base);
    PlayMusicStream(game.resources.bgm.intensity);
    PlayMusicStream(game.resources.bgm.miniboss);
    PlayMusicStream(game.resources.bgm.megaboss);
    PlayMusicStream(game.resources.bgm.swarmBoss);
    SetMusicVolume(game.resources.bgm.base, SoundConstants::baseVolume);
    SetMusicVolume(game.resources.bgm.intensity, 0);
    SetMusicVolume(game.resources.bgm.miniboss, 0);
    SetMusicVolume(game.resources.bgm.megaboss, 0);
    SetMusicVolume(game.resources.bgm.swarmBoss, 0);

    game.resources.renderScale = computeRenderScale(game);
    createRenderTargets(game);

    resetRun(game);
    game.state = GameState::TITLE;

    return game;
}

void toggleFullscreen(Game& game)
{
    if (IsWindowFullscreen())
    {
        ToggleFullscreen();
        SetWindowSize(GameConstants::defaultWindowWidth, GameConstants::defaultWindowHeight);
    }
    else
    {
        const int monitor = GetCurrentMonitor();
        SetWindowSize(GetMonitorWidth(monitor), GetMonitorHeight(monitor));
        ToggleFullscreen();
    }

    syncScreenSize(game);
}

void syncScreenSize(Game& game)
{
    game.resources.windowWidth = GetScreenWidth();
    game.resources.windowHeight = GetScreenHeight();

    if (const float newScale = computeRenderScale(game); newScale != game.resources.renderScale)
    {
        game.resources.renderScale = newScale;
        createRenderTargets(game);
    }
}

void resetRun(Game& game)
{
    const ShipDef& ship = currentShip(game);

    game.run.player = Player{.position = Vector2{},
                             .velocity = Vector2{},
                             .radius = ship.radius,
                             .color = ship.color,
                             .speed = ship.speed,
                             .health = ship.maxHealth,
                             .maxHealth = ship.maxHealth,
                             .shieldActive = false,
                             .shieldTimer = 0,
                             .shieldCooldownTimer = 0,
                             .immunityTimer = 0,
                             .blackHoleCoreTimer = 0,
                             .bossBodyTimer = 0,
                             .slowTimer = 0,
                             .charges = UpdateConstants::maxCharges,
                             .chargeRegenTimer = 0,
                             .dashing = false,
                             .dashTimer = 0,
                             .dashVelocity = Vector2{},
                             .shieldStacks = 0,
                             .nerve = 0,
                             .nerveCharging = false,
                             .nerveChargeTimer = 0,
                             .nerveChargeFeedTimer = 0,
                             .elementalBuffTimer = {},
                             .regenTimer = 0,
                             .regenRate = 0,
                             .overchargeTimer = 0,
                             .dashTrailTimer = 0,
                             .secondWindReady = false,
                             .overdriveTimer = 0};

    game.run.bosses.clear();

    game.run.bullets.clear();
    game.run.asteroids.clear();

    game.run.asteroids.reserve(static_cast<size_t>(maxAsteroid));
    game.run.bossProjectiles.clear();
    game.run.enemies.clear();

    game.run.enemies.reserve(static_cast<size_t>(UpdateConstants::maxEnemies));
    game.run.eliteHazards.clear();
    game.run.eliteHazardSpawnTimer = static_cast<float>(GetRandomValue(45, 90));
    game.run.pickups.clear();
    game.run.mines.clear();
    game.run.orbitBladeProjectiles.clear();
    game.run.nerveBallProjectiles.clear();
    game.run.nerveSpiralProjectiles.clear();
    game.run.deathParticles.clear();
    game.run.dashTrailParticles.clear();
    game.run.flameParticles.clear();
    game.run.damageNumbers.clear();
    game.run.weaponDowngrade = std::nullopt;
    game.run.followerDrones.clear();
    game.run.laserDrones.clear();
    game.run.turrets.clear();
    game.run.chainLightningBolts.clear();
    game.run.sniperShots.clear();
    game.run.weapons = {Weapon{.type = ship.defaultWeapon, .level = 1}};
    game.run.skillLevels.fill(0);
    game.run.skillLevels.at(
        static_cast<size_t>(weaponGrantSkill.at(static_cast<size_t>(ship.defaultWeapon)))) = 1;
    game.run.postCapDamageLevels = 0;
    game.run.damageMeter = DamageMeter{};
    game.run.damageMeterDisplay = DamageMeter{};
    game.run.damageMeterHoldTimer = 0;
    game.run.bossDeathShockwaves.clear();

    game.run.blackhole = BlackHole{};
    game.run.blackhole.timer = static_cast<float>(GetRandomValue(30, 60)) / 10.0F;
    game.run.wormhole = Wormhole{};
    game.run.wormhole.timer = static_cast<float>(GetRandomValue(400, 700)) / 10.0F;
    game.run.asteroidSpawnTimer = 1.0F;
    game.run.enemySpawnTimer = 1.0F;
    game.run.solarForgeHeatMeter = 0;
    game.run.solarForgeHeatTickTimer = 0;
    game.run.fluidContactTickTimer = 0;
    game.run.organicMergeTimer = organicMergeCheckInterval;

    game.run.xp = 0;
    game.run.level = 1;
    game.run.xpToNext = 100;
    game.run.waveNumber = 1;
    game.run.waveTimer = GameConstants::waveDuration;
    game.run.bossSpawnCount = 0;
    game.run.runTime = 0;

    game.run.shakeTimer = 0;
    game.run.shakeDuration = 0;
    game.run.shakeIntensity = 0;
    game.run.hitPauseTimer = 0;
    game.run.nerveBurstFlashTimer = 0;
    game.run.nerveBurstFlashEnd = Vector2{};
    game.run.score = 0;
    game.run.scoreRecorded = false;
    game.run.achievementToast.clear();
    game.run.achievementToastTimer = 0;
    game.state = GameState::GAMEPLAY;
}
